pest/src/inputs_private/position.rs

```diff
@@ -0,0 +1,520 @@
+// pest. The Elegant Parser
+// Copyright (C) 2017  Dragoș Tiselice
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::cmp::Ordering;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::Range;
+use std::rc::Rc;
+
+use super::input::Input;
+use super::span;
+
+/// A `struct` containing a position that is tied to an `Input` which provides useful methods to
+/// manually parse it. This leads to an API largely based on the standard `Result`.
+pub struct Position<I: Input> {
+    input: Rc<I>,
+    pos: usize
+}
+
+pub fn new<I: Input>(input: Rc<I>, pos: usize) -> Position<I> {
+    Position {
+        input: input,
+        pos: pos
+    }
+}
+
+impl<I: Input> Position<I> {
+    /// Creates starting `Position` from an `Rc<Input>`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::rc::Rc;
+    /// # use pest::inputs::{Position, StringInput};
+    /// let input = Rc::new(StringInput::new("".to_owned()));
+    ///
+    /// Position::from_start(input);
+    /// ```
+    #[inline]
+    pub fn from_start(input: Rc<I>) -> Position<I> {
+        Position {
+            input: input,
+            pos: 0
+        }
+    }
+
+    /// Returns the current position as a `usize`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::rc::Rc;
+    /// # use pest::inputs::{Position, StringInput};
+    /// let input = Rc::new(StringInput::new("ab".to_owned()));
+    /// let start = Position::from_start(input);
+    ///
+    /// assert_eq!(start.pos(), 0);
+    /// assert_eq!(start.match_string("ab").unwrap().pos(), 2);
+    /// ```
+    #[inline]
+    pub fn pos(&self) -> usize {
+        self.pos
+    }
+
+    /// Creates a `Span` from two `Position`s.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::rc::Rc;
+    /// # use pest::inputs::{Position, StringInput};
+    /// let input = Rc::new(StringInput::new("ab".to_owned()));
+    /// let start = Position::from_start(input);
+    /// let end = start.clone().match_string("ab").unwrap();
+    /// let span = start.span(end);
+    ///
+    /// assert_eq!(span.start(), 0);
+    /// assert_eq!(span.end(), 2);
+    /// ```
+    #[inline]
+    pub fn span(self, other: Position<I>) -> span::Span<I> {
+        if &*self.input as *const I == &*other.input as *const I {
+            span::new(self.input, self.pos, other.pos)
+        } else {
+            panic!("Span created from positions from different inputs")
```

Comment:
I would argue that having multiple `Input`s and actually hitting this panic is a rare event which is why I'd keep the syntax simpler here and not wrap in an `Option` or `Result`. What do you think?